File: js/src/program/ops/opObjectConstructor.js
import { JPLTypeError } from '../../library';
import { call } from './utils';

export default {
  /** { fields: [{ key: [op], value: [op], optional: boolean }] } */
  async op(runtime, input, params, scope, next) {
    const fields = await runtime.muxAsync([params.fields], async (field) => {
      const [keys, values] = await Promise.all([
        runtime.executeInstructions(field.key, [input], scope),
        runtime.executeInstructions(field.value, [input], scope),
      ]);

      return runtime.muxAll([runtime.unwrapValues(keys), values], (key, value) => {
        const t = runtime.type(key);
        switch (t) {
          case 'string':
            return [[key, value]];

          default:
        }

        if (field.optional) return [];
        throw new JPLTypeError('cannot use %s (%*<100v) as object key', t, key);
      });
    });

    return runtime.muxAll(fields, (...entries) => next(Object.fromEntries(entries), scope));
  },

  /** { fields: [{ key: function, value: function, optional: boolean }] } */
  map(runtime, params) {
    return {
      fields: runtime.muxOne([params.fields], (entry) => ({
        key: call(entry.key),
        value: call(entry.value),
        optional: runtime.assertType(entry.optional, 'boolean'),
      })),
    };
  },
};

File: go/program/ops/opObjectConstructor.go
package ops

import (
	"github.com/2manyvcos/jpl/go/definition"
	"github.com/2manyvcos/jpl/go/jpl"
	"github.com/2manyvcos/jpl/go/library"
)

type opObjectConstructor struct{}

// { fields: [{ key: [op], value: [op], optional: boolean }] }
func (opObjectConstructor) OP(runtime jpl.JPLRuntime, input any, params definition.JPLInstructionParams, scope jpl.JPLRuntimeScope, next jpl.JPLScopedPiper) ([]any, jpl.JPLError) {
	fields, err := library.MuxOne([][]definition.JPLField{params.Fields}, jpl.IOMuxerFunc[definition.JPLField, []*library.ObjectEntry[any]](func(args ...definition.JPLField) ([]*library.ObjectEntry[any], jpl.JPLError) {
		field := args[0]
		keys, err := runtime.ExecuteInstructions(field.Key, []any{input}, scope, nil)
		if err != nil {
			return nil, err
		}
		values, err := runtime.ExecuteInstructions(field.Value, []any{input}, scope, nil)
		if err != nil {
			return nil, err
		}

		unwrappedKeys, err := library.UnwrapValues(keys, "")
		if err != nil {
			return nil, err
		}
		return library.MuxAll([][]any{unwrappedKeys, values}, jpl.IOMuxerFunc[any, []*library.ObjectEntry[any]](func(args ...any) ([]*library.ObjectEntry[any], jpl.JPLError) {
			key := args[0]
			value := args[1]
			t, err := library.Type(key)
			if err != nil {
				return nil, err
			}
			switch t {
			case jpl.JPLT_STRING:
				return []*library.ObjectEntry[any]{{Key: key.(string), Value: value}}, nil

			default:
			}

			if field.Optional {
				return nil, nil
			}
			return nil, library.ThrowAny(library.NewTypeError("cannot use %s (%*<100v) as object key", string(t), key))
		}))
	}))
	if err != nil {
		return nil, err
	}

	return library.MuxAll(fields, jpl.IOMuxerFunc[*library.ObjectEntry[any], []any](func(entries ...*library.ObjectEntry[any]) ([]any, jpl.JPLError) {
		return next.Pipe(library.ObjectFromEntries(entries), scope)
	}))
}

// { fields: [{ key: function, value: function, optional: boolean }] }
func (opObjectConstructor) Map(runtime jpl.JPLRuntime, params jpl.JPLInstructionParams) (result definition.JPLInstructionParams, err jpl.JPLError) {
	if result.Fields, err = library.MuxOne([][]jpl.JPLField{params.Fields}, jpl.IOMuxerFunc[jpl.JPLField, definition.JPLField](func(args ...jpl.JPLField) (definition.JPLField, jpl.JPLError) {
		entry := args[0]
		return definition.JPLField{
			Key:      call(entry.Key),
			Value:    call(entry.Value),
			Optional: entry.Optional,
		}, nil
	})); err != nil {
		return
	}
	return
}
